### PR TITLE
Issue #6687 Fix QueryBuilder filtering for AbstractCode

### PR DESCRIPTION
**Pull request by adityagh006**
This PR fixes an issue in QueryBuilder where filtering for AbstractCode was not working correctly. The update ensures that queries involving AbstractCode return expected results.

Changes Made:
1.Fixed filtering logic in querybuilder.py.
2.Updated pyproject.toml accordingly.

Let me know if any additional changes are needed.

### CODE REVIEW

File: src/aiida/orm/querybuilder.py [hunk 1]
```diff
@@ -1336,13 +1336,23 @@ def _get_node_type_filter(classifiers: Classifier, subclassing: bool) -> dict:
...
 
+    # value = classifiers.ormclass_type_string
+
+    # if not subclassing:
+    #     filters = {'==': value}
+    # else:
+    #     # Note: the query_type_string always ends with a dot. This ensures that "like {str}%" matches *only*
+    #     # the query type string
+    #     filters = {'like': f'{escape_for_sql_like(get_query_type_from_type_string(value))}%'}
+
```
Comment (danielhollas) [inline src/aiida/orm/querybuilder.py:1347]: Please remove all commented out code (also below)

```suggestion
```

File: src/aiida/orm/querybuilder.py [hunk 2]
```diff
@@ -1336,13 +1336,23 @@ def _get_node_type_filter(classifiers: Classifier, subclassing: bool) -> dict:
...
     value = classifiers.ormclass_type_string
 
+    if value == 'data.core.code.abstract':
```
Comment (danielhollas) [inline src/aiida/orm/querybuilder.py:1350]: I think this needs a longer comment to explain this corner case. You can perhaps reuse part of the explanation from the issue.

**Review by danielhollas — CHANGES_REQUESTED** [review]
Thank you for the contribution @adityagh006! :clap: 

This will need tests, which should cover various types of codes as described in the issue. Hopefully the existing QueryBuilder tests in `aiida-core/tests/orm/test_querybuilder.py` can help you get started. Let us now if you need assistance!

File: src/aiida/orm/querybuilder.py (hunk 1 above)
Comment (adityagh006) [reply src/aiida/orm/querybuilder.py:1347]: Oh, sorry for that! I forgot to remove those comments. I'll take care of it in the next update.

File: src/aiida/orm/querybuilder.py (hunk 2 above)
Comment (adityagh006) [reply src/aiida/orm/querybuilder.py:1350]: Got it! I'll add a detailed comment explaining this corner case, incorporating relevant details from the issue.

**Comment by danielhollas** [comment]
Hi @adityagh006, are you still planning to work on this PR? I think the code looks good (though I haven't tested it personally) but it needs unit tests.

**Comment by adityagh006** [comment]
@danielhollas! I’ve written some test cases for the AbstractCode filtering, but I’m having trouble getting them to run properly. I’ve tried a few things, but I'm not sure if it’s an environment issue or something else—I’m a bit stuck at the moment.

**Comment by danielhollas** [comment]
@adityagh006 no worries. Feel free to push what you have and I can have a look.

File: tests/orm/test_querybuilder.py [hunk 3]
```diff
@@ -852,6 +852,59 @@ def test_empty_filters(self):
...
+        # Create test codes of different types
+        installed_code = orm.InstalledCode(
+            label='test_installed_abstract',
+            computer=computer,
+            filepath_executable='/bin/bash',
+        ).store()
+
+        portable_code = orm.PortableCode(
+            label='test_portable_abstract',
+            filepath_executable='/bin/bash',
```
Comment (danielhollas) [inline tests/orm/test_querybuilder.py:880]: I think this needs to be fixed, please look at existing tests at tests/orm/data/code/test_portable.py
```suggestion
            filepath_paths='/bin/bash',
```

File: tests/orm/test_querybuilder.py [hunk 4]
```diff
@@ -852,6 +852,59 @@ def test_empty_filters(self):
...
+    from aiida.orm import AbstractCode, Computer, InstalledCode, PortableCode, QueryBuilder
+
+    def test_abstract_code_filtering(self):
+        """Test that querying for AbstractCode correctly returns all code instances.
+
+        This tests the fix for issue #6687, where QueryBuilder couldn't find codes
+        when looking for AbstractCode due to a node_type mismatch.
+        """
+        # Set up test environment
+        computer = orm.Computer(
```
Comment (danielhollas) [inline tests/orm/test_querybuilder.py:864]: Please use an existing fixture [aiida_computer_local](https://aiida.readthedocs.io/projects/aiida-core/en/stable/topics/plugins.html#aiida-computer-local) to simplify this. You can then delete the cleanup code since the fixture does that automatically at the end of the test.

File: tests/orm/test_querybuilder.py [hunk 5]
```diff
@@ -852,6 +852,59 @@ def test_empty_filters(self):
         qb = orm.QueryBuilder().append(orm.Data, filters={'or': [{}, {}]})
         assert qb.count() == count
 
+    from aiida.orm import AbstractCode, Computer, InstalledCode, PortableCode, QueryBuilder
```
Comment (danielhollas) [inline tests/orm/test_querybuilder.py:855]: Let's move these imports at the top of the file

File: tests/orm/test_querybuilder.py [hunk 6]
```diff
@@ -852,6 +852,59 @@ def test_empty_filters(self):
...
+        portable_code = orm.PortableCode(
+            label='test_portable_abstract',
+            filepath_executable='/bin/bash',
+        ).store()
+
+        try:
+            # Test 1: AbstractCode query should find all code types
+            qb_abstract = orm.QueryBuilder().append(orm.AbstractCode)
+            abstract_results = qb_abstract.all(flat=True)
+            assert installed_code in abstract_results, 'InstalledCode not found with AbstractCode query'
```
Comment (danielhollas) [inline tests/orm/test_querybuilder.py:887]: I don't think these asserts will work, since the objects returned by the QueryBuilder will be different from the original objects that you stored, even though they point to the same DB node. I think you should compare the `id` of the nodes, something like

```suggestion
            abstract_ids = [code.id for code in abstract_results]
            assert installed_code.id in abstract_ids, 'InstalledCode not found with AbstractCode query'
```

**Review by danielhollas — CHANGES_REQUESTED** [review]
@adityagh006 can you please rebase this PR on latest main? It now contains some extra commits.

The logic of the new tests looks good! Just needs a couple of fixes, see my comments. Let me know if you need further help, and thanks for working through this!

File: pyproject.toml [hunk 7]
```diff
@@ -137,6 +138,9 @@ requires-python = '>=3.9'
 'process.workflow.workchain' = 'aiida.orm.nodes.process.workflow.workchain:WorkChainNode'
 'process.workflow.workfunction' = 'aiida.orm.nodes.process.workflow.workfunction:WorkFunctionNode'
 
+[project.entry-points.'aiida.orm']
+'core.code.abstract' = 'aiida.orm.nodes.data.code.abstract:AbstractCode'
```
Comment (danielhollas) [inline pyproject.toml:142]: I think this entry point should be moved to `[project.entry-points.'aiida.data']` section where the other code entry points are defined.

File: pyproject.toml [hunk 8]
```diff
@@ -183,6 +187,7 @@ requires-python = '>=3.9'
 [project.entry-points.'aiida.workflows']
 'core.arithmetic.add_multiply' = 'aiida.workflows.arithmetic.add_multiply:add_multiply'
 'core.arithmetic.multiply_add' = 'aiida.workflows.arithmetic.multiply_add:MultiplyAddWorkChain'
+'core.code.abstract' = 'aiida.orm.nodes.data.code.abstract:AbstractCode'
```
Comment (danielhollas) [inline pyproject.toml:190]:
```suggestion
```

**Comment by adityagh006** [comment]
Hi @danielhollas,
I’ve been struggling to rebase the branch and push the changes due to some virtual environment issues. As a result, I mistakenly force-pushed, which caused the PR to close.
Could you please let me know what the next steps are or if there’s anything I should do to reopen or proceed with this PR?

Apologies for the inconvenience.

**Comment by danielhollas** [comment]
@adityagh006 feel free to open a new one. 

I can help with the rebase, but there are currently no commits in this branch so I cannot...

**Comment by adityagh006** [comment]
Hi @danielhollas,

Apologies for the delay, but I’ve created a new PR for the changes. You can find it [here](https://github.com/aiidateam/aiida-core/pull/6857).

Let me know if any further adjustments are needed.

Thanks for your help!

**Comment by adityagh006** [comment]
Hi @danielhollas,

Apologies for the delay, but I’ve created a new PR for the changes. You can find it [here](https://github.com/aiidateam/aiida-core/pull/6857).

Let me know if any further adjustments are needed.
